File: ios/Classes/Models/HMSPeerExtension.swift
//
//  HMSPeerExtension.swift
//  hmssdk_flutter
//
//  Copyright © 2021 100ms. All rights reserved.
//

import Foundation
import HMSSDK

class  HMSPeerExtension {

    static func toDictionary(_ peer: HMSPeer) -> [String: Any] {

        var dict = [
            "peer_id": peer.peerID,
            "name": peer.name,
            "is_local": peer.isLocal,
            "customer_description": peer.metadata ?? "",
            "customer_user_id": peer.customerUserID ?? "",
            "joined_at": "\(peer.joinedAt)",
            "updated_at": "\(peer.updatedAt)"
        ] as [String: Any]

        if let metadata = peer.metadata {
            dict["metadata"] = metadata
        }
        if let role = peer.role {
            dict["role"] = HMSRoleExtension.toDictionary(role)
        }

        if let audioTrack = peer.audioTrack {
            dict["audio_track"] = HMSTrackExtension.toDictionary(audioTrack)
        }

        if let videoTrack = peer.videoTrack {
            dict["video_track"] = HMSTrackExtension.toDictionary(videoTrack)
        }

        if let aux = peer.auxiliaryTracks {
            var auxilaryTracks = [[String: Any]]()
            aux.forEach { auxilaryTracks.append(HMSTrackExtension.toDictionary($0)) }
            dict["auxilary_tracks"] = auxilaryTracks
        }

        if let quality = peer.networkQuality {
            dict["network_quality"] = HMSNetworkQualityExtension.toDictionary(quality)
        }
        return dict
    }

    static func getValueOf(_ update: HMSPeerUpdate) -> String {
        switch update {
        case .peerJoined:
            return "peerJoined"
        case .peerLeft:
            return "peerLeft"
        case .roleUpdated:
            return "roleUpdated"
        case .defaultUpdate:
            return "defaultUpdate"
        case .nameUpdated:
            return "nameChanged"
        case .metadataUpdated:
            return "metadataChanged"
        case .networkQualityUpdated:
            return "networkQualityUpdated"
        @unknown default:
            return "defaultUpdate"
        }
    }
}

File: android/src/main/kotlin/live/hms/hmssdk_flutter/HMSPeerExtension.kt
package live.hms.hmssdk_flutter

import live.hms.video.sdk.models.HMSPeer
import live.hms.video.sdk.models.enums.HMSPeerUpdate
import java.text.SimpleDateFormat

class HMSPeerExtension {

    companion object {
        fun toDictionary(peer: HMSPeer?): HashMap<String, Any?>? {
            val args = HashMap<String, Any?>()

            if (peer == null)return null
            args["peer_id"] = peer.peerID
            args["name"] = peer.name
            args["is_local"] = peer.isLocal
            args["role"] = HMSRoleExtension.toDictionary(peer.hmsRole)
            args["metadata"] = peer.metadata
            args["customer_user_id"] = peer.customerUserID
            args["audio_track"] = HMSTrackExtension.toDictionary(peer.audioTrack)
            args["video_track"] = HMSTrackExtension.toDictionary(peer.videoTrack)
            args["network_quality"] = HMSNetworkQualityExtension.toDictionary(peer.networkQuality)
            args["joined_at"] = SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(peer.joinedAt).toString()

            val auxTrackList = ArrayList<Any>()
            peer.auxiliaryTracks.forEach {
                auxTrackList.add(HMSTrackExtension.toDictionary(it)!!)
            }
            args["auxilary_tracks"] = auxTrackList
            return args
        }

        fun getValueofHMSPeerUpdate(update: HMSPeerUpdate?): String? {
            if (update == null)return null

            return when (update) {
                HMSPeerUpdate.PEER_JOINED -> "peerJoined"
                HMSPeerUpdate.PEER_LEFT -> "peerLeft"
                HMSPeerUpdate.ROLE_CHANGED -> "roleUpdated"
                HMSPeerUpdate.METADATA_CHANGED -> "metadataChanged"
                HMSPeerUpdate.NAME_CHANGED -> "nameChanged"
                HMSPeerUpdate.NETWORK_QUALITY_UPDATED -> "networkQualityUpdated"
                HMSPeerUpdate.BECAME_DOMINANT_SPEAKER -> "becameDominantSpeaker"
                HMSPeerUpdate.NO_DOMINANT_SPEAKER -> "noDominantSpeaker"
                else -> "defaultUpdate"
            }
        }
    }
}
